File: rustlox/src/scanner.rs
use std::iter::{Enumerate, Peekable};

#[derive(Copy, Clone, PartialEq)]
#[repr(u8)]
pub enum TokenKind {
    // Single-character tokens.
    LeftParen,
    RightParen,
    LeftBrace,
    RightBrace,
    Comma,
    Dot,
    Minus,
    Plus,
    Semicolon,
    Slash,
    Star,
    // One or two character tokens.
    Bang,
    BangEqual,
    Equal,
    EqualEqual,
    Greater,
    GreaterEqual,
    Less,
    LessEqual,
    // Literals.
    Identifier,
    String,
    Number,
    // Keywords.
    And,
    Class,
    Else,
    False,
    For,
    Fun,
    If,
    Nil,
    Or,
    Print,
    Return,
    Super,
    This,
    True,
    Var,
    While,

    Error,
}

#[derive(Copy, Clone)]
pub struct Token<'a> {
    pub kind: TokenKind,
    pub line: i32,
    pub lexeme: &'a str,
}

pub struct Scanner<'a> {
    source: &'a String,
    pub lines: i32,
    start: usize,
    iter: Peekable<Enumerate<std::str::Chars<'a>>>,
}

impl<'a> Scanner<'a> {
    pub fn new(source: &'a String) -> Scanner<'a> {
        Scanner {
            source,
            lines: 1,
            start: 0,
            iter: source.chars().enumerate().peekable(),
        }
    }

    fn advance(&mut self) -> Option<(usize, char)> {
        self.iter.next()
    }

    fn consume_while(&mut self, fun: fn(c: char) -> bool) {
        while self.iter.next_if(|&(_, c)| fun(c)).is_some() {}
    }

    fn match_current(&mut self, expected: char) -> bool {
        self.iter.next_if(|&(_, c)| c == expected).is_some()
    }

    fn skip_whitespace(&mut self) {
        while let Some((_, c)) = self.iter.peek() {
            match *c {
                ' ' | '\r' | '\t' => {
                    self.advance();
                }
                '\n' => {
                    self.lines += 1;
                    self.advance();
                }
                '/' => {
                    if let Some((_, '/')) = self.peek_next() {
                        self.consume_while(|c| c != '\n');
                    } else {
                        return;
                    }
                }
                _ => return,
            }
        }
    }

    fn get_lexeme(&mut self) -> &'a str {
        let end = if let Some((number, _)) = self.iter.peek() {
            *number
        } else {
            self.source.len()
        };

        &self.source[self.start..end]
    }

    fn make_token(&mut self, kind: TokenKind) -> Token<'a> {
        Token {
            kind: kind,
            line: self.lines,
            lexeme: self.get_lexeme(),
        }
    }

    fn make_error_token(&self, message: &'static str) -> Token<'a> {
        Token {
            kind: TokenKind::Error,
            line: self.lines,
            lexeme: message,
        }
    }

    fn peek_next(&mut self) -> Option<(usize, char)> {
        if let Some((n, _)) = self.iter.peek() {
            if let Some(byte) = self.source.as_bytes().get(n + 1) {
                return Some((n + 1, *byte as char));
            }
        }

        None
    }

    fn string(&mut self) -> Token<'a> {
        while let Some((_, c)) = self.iter.next_if(|&(_, c)| c != '"') {
            if c == '\n' {
                self.lines += 1;
            }
        }

        if self.match_current('"') {
            self.make_token(TokenKind::String)
        } else {
            self.make_error_token("Unterminated string.")
        }
    }

    fn number(&mut self) -> Token<'a> {
        self.consume_while(|c| c.is_digit(10));

        // Look for a fractional part.
        if let Some((_, '.')) = self.iter.peek() {
            if let Some((_, '0'..='9')) = self.peek_next() {
                // Consume the ".".
                self.advance();
                self.consume_while(|c| c.is_digit(10));
            }
        }

        self.make_token(TokenKind::Number)
    }

    fn identifier(&mut self) -> Token<'a> {
        self.consume_while(|c| c.is_ascii_alphanumeric() || c == '_');

        let lexeme = self.get_lexeme();
        let kind = match lexeme {
            "and" => TokenKind::And,
            "class" => TokenKind::Class,
            "else" => TokenKind::Else,
            "false" => TokenKind::False,
            "for" => TokenKind::For,
            "fun" => TokenKind::Fun,
            "if" => TokenKind::If,
            "nil" => TokenKind::Nil,
            "or" => TokenKind::Or,
            "print" => TokenKind::Print,
            "return" => TokenKind::Return,
            "super" => TokenKind::Super,
            "this" => TokenKind::This,
            "true" => TokenKind::True,
            "var" => TokenKind::Var,
            "while" => TokenKind::While,
            _ => TokenKind::Identifier,
        };

        Token {
            kind: kind,
            lexeme: lexeme,
            line: self.lines,
        }
    }
}

impl<'a> Iterator for Scanner<'a> {
    type Item = Token<'a>;

    fn next(&mut self) -> Option<Token<'a>> {
        self.skip_whitespace();

        let next = self.advance();

        if next.is_none() {
            return None;
        }

        let (start, c) = next.unwrap();
        self.start = start;

        let token = match c {
            '(' => self.make_token(TokenKind::LeftParen),
            ')' => self.make_token(TokenKind::RightParen),
            '{' => self.make_token(TokenKind::LeftBrace),
            '}' => self.make_token(TokenKind::RightBrace),
            ';' => self.make_token(TokenKind::Semicolon),
            ',' => self.make_token(TokenKind::Comma),
            '.' => self.make_token(TokenKind::Dot),
            '-' => self.make_token(TokenKind::Minus),
            '+' => self.make_token(TokenKind::Plus),
            '*' => self.make_token(TokenKind::Star),
            '/' => self.make_token(TokenKind::Slash),
            '!' => {
                if self.match_current('=') {
                    self.make_token(TokenKind::BangEqual)
                } else {
                    self.make_token(TokenKind::Bang)
                }
            }
            '=' => {
                if self.match_current('=') {
                    self.make_token(TokenKind::EqualEqual)
                } else {
                    self.make_token(TokenKind::Equal)
                }
            }
            '<' => {
                if self.match_current('=') {
                    self.make_token(TokenKind::LessEqual)
                } else {
                    self.make_token(TokenKind::Less)
                }
            }
            '>' => {
                if self.match_current('=') {
                    self.make_token(TokenKind::GreaterEqual)
                } else {
                    self.make_token(TokenKind::Greater)
                }
            }
            '"' => self.string(),
            '0'..='9' => self.number(),
            'a'..='z' | 'A'..='Z' | '_' => self.identifier(),
            _ => self.make_error_token("Unexpected character."),
        };

        Some(token)
    }
}

File: tslox/src/Scanner.ts
import type { LoxError } from "./main.ts";
import Token from "./Token.ts";
import TokenType from "./TokenType.ts";

export default class Scanner {
  static keywords: Record<string, TokenType> = Object.assign(
    Object.create(null),
    {
      and: TokenType.And,
      break: TokenType.Break,
      class: TokenType.Class,
      else: TokenType.Else,
      false: TokenType.False,
      for: TokenType.For,
      fun: TokenType.Fun,
      if: TokenType.If,
      nil: TokenType.Nil,
      or: TokenType.Or,
      print: TokenType.Print,
      return: TokenType.Return,
      super: TokenType.Super,
      this: TokenType.This,
      true: TokenType.True,
      var: TokenType.Var,
      while: TokenType.While,
    },
  );

  private static isDigit(char: string): boolean {
    return char >= "0" && char <= "9";
  }

  private static isAlpha(char: string): boolean {
    return (char >= "a" && char <= "z") || (char >= "A" && char <= "Z") ||
      char === "_";
  }

  private static isAlphaNumeric(char: string): boolean {
    return this.isDigit(char) || this.isAlpha(char);
  }

  private start = 0;
  private current = 0;
  private line = 1;

  constructor(
    private readonly source: string,
    private readonly loxError: LoxError,
  ) {}

  private advance(): string {
    return this.source[this.current++]!;
  }

  private createToken(type: TokenType): Token;
  private createToken(type: TokenType.String, literal: string): Token;
  private createToken(type: TokenType.Number, literal: number): Token;
  private createToken(type: TokenType, literal: any = null): Token {
    const text = this.source.slice(this.start, this.current);
    return new Token(type, text, literal, this.line);
  }

  private match(expected: string): boolean {
    if (this.isAtEnd()) return false;
    if (this.source[this.current] !== expected) return false;

    this.current++;
    return true;
  }

  private peek(): string {
    return this.source[this.current] ?? "\0";
  }

  private peekNext(): string {
    return this.source[this.current + 1] ?? "\0";
  }

  private string(): Token | null {
    while (this.peek() !== '"' && !this.isAtEnd()) {
      if (this.peek() === "\n") this.line++;
      this.advance();
    }

    if (this.isAtEnd()) {
      this.loxError(this.line, "Unterminated string.");
      return null;
    }

    // consume second "
    this.advance();

    const value = this.source.slice(this.start + 1, this.current - 1);
    return this.createToken(TokenType.String, value);
  }

  private number(): Token {
    while (Scanner.isDigit(this.peek())) {
      this.advance();
    }

    if (this.peek() === "." && Scanner.isDigit(this.peekNext())) {
      // consume the .
      this.advance();

      while (Scanner.isDigit(this.peek())) {
        this.advance();
      }
    }

    return this.createToken(
      TokenType.Number,
      Number.parseFloat(this.source.slice(this.start, this.current)),
    );
  }

  private identifier(): Token {
    while (Scanner.isAlphaNumeric(this.peek())) {
      this.advance();
    }

    const text = this.source.slice(this.start, this.current);
    const type = Scanner.keywords[text] ?? TokenType.Identifier;
    return this.createToken(type);
  }

  private scanToken(): Token | null {
    const char = this.advance();
    switch (char) {
      case "(":
        return this.createToken(TokenType.LeftParen);
      case ")":
        return this.createToken(TokenType.RightParen);
      case "{":
        return this.createToken(TokenType.LeftBrace);
      case "}":
        return this.createToken(TokenType.RightBrace);
      case ",":
        return this.createToken(TokenType.Comma);
      case ".":
        return this.createToken(TokenType.Dot);
      case "-":
        return this.createToken(TokenType.Minus);
      case "+":
        return this.createToken(TokenType.Plus);
      case ";":
        return this.createToken(TokenType.Semicolon);
      case "*":
        return this.createToken(TokenType.Star);
      case "?":
        return this.createToken(TokenType.QuestionMark);
      case ":":
        return this.createToken(TokenType.Colon);
      case "!":
        return this.createToken(
          this.match("=") ? TokenType.BangEqual : TokenType.Bang,
        );
      case "=":
        return this.createToken(
          this.match("=") ? TokenType.EqualEqual : TokenType.Equal,
        );
      case "<":
        return this.createToken(
          this.match("=") ? TokenType.LessEqual : TokenType.Less,
        );
      case ">":
        return this.createToken(
          this.match("=") ? TokenType.GreaterEqual : TokenType.Greater,
        );
      case "/":
        if (this.match("/")) {
          while (this.peek() !== "\n" && !this.isAtEnd()) this.advance();
        } else if (this.match("*")) {
          while (this.peek() !== "*" && !this.match("/") && !this.isAtEnd()) {
            const current = this.advance();
            if (current === "\n") this.line++;
          }
          this.advance();
          this.advance();
        } else {
          return this.createToken(TokenType.Slash);
        }
        break;
      case " ":
      case "\r":
      case "\t":
        break;
      case "\n":
        this.line++;
        break;
      case '"':
        return this.string();
      default:
        if (Scanner.isDigit(char)) return this.number();
        if (Scanner.isAlpha(char)) return this.identifier();
        this.loxError(this.line, "Unexpected character.");
    }
    return null;
  }

  private isAtEnd(): boolean {
    return this.current >= this.source.length;
  }

  *[Symbol.iterator](): IterableIterator<Token> {
    while (!this.isAtEnd()) {
      this.start = this.current;
      const token = this.scanToken();
      if (token) yield token;
    }

    yield new Token(TokenType.EOF, "", null, this.line);
  }
}
